staging/src/k8s.io/sample-controller/integration_test.go

```diff
@@ -0,0 +1,215 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	extensionclientset "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	exttesting "k8s.io/apiextensions-apiserver/test/integration/testserver"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	kubeinformers "k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	apitesting "k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
+	"k8s.io/kubernetes/test/integration/framework"
+	clientset "k8s.io/sample-controller/pkg/client/clientset/versioned"
+	informers "k8s.io/sample-controller/pkg/client/informers/externalversions"
+)
+
+func TestMain(m *testing.M) {
+	framework.EtcdMain(m.Run)
+}
+
+func TestController(t *testing.T) {
+	// Start an API server
+	result := apitesting.StartTestServerOrDie(t, nil, framework.SharedEtcd())
+	defer result.TearDownFn()
+
+	// Install the CRD
+	// TODO: Load this from artifacts/examples/crd.yaml
+	apiExtensionClient, err := extensionclientset.NewForConfig(result.ClientConfig)
+	if err != nil {
+		t.Fatalf("error creating extension clientset: %v", err)
+	}
+	crd := &apiextensionsv1beta1.CustomResourceDefinition{
+		ObjectMeta: metav1.ObjectMeta{Name: "foos.samplecontroller.k8s.io"},
+		Spec: apiextensionsv1beta1.CustomResourceDefinitionSpec{
+			Group:   "samplecontroller.k8s.io",
+			Version: "v1alpha1",
+			Names: apiextensionsv1beta1.CustomResourceDefinitionNames{
+				Plural:   "foos",
+				Singular: "foo",
+				Kind:     "Foo",
+			},
+			Scope: apiextensionsv1beta1.NamespaceScoped,
+		},
+	}
+
+	pool := dynamic.NewDynamicClientPool(result.ClientConfig)
+	dynamicClient, err := exttesting.CreateNewCustomResourceDefinition(crd, apiExtensionClient, pool)
+	if err != nil {
+		t.Fatalf("error while creating CRD: %v", err)
+	}
+
+	// Create clients
+	kubeClient, err := kubernetes.NewForConfig(result.ClientConfig)
+	if err != nil {
+		t.Fatalf("error creating kubernetes clientset: %v", err.Error())
+	}
+
+	exampleClient, err := clientset.NewForConfig(result.ClientConfig)
+	if err != nil {
+		t.Fatalf("error creating clientset: %v", err)
+	}
+
+	// Create a test namespace
+	ns := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "namespace-for-test",
+		},
+	}
+	_, err = kubeClient.Core().Namespaces().Create(ns)
+	if err != nil {
+		t.Fatalf("error creating namespace: %v", err)
+	}
+
+	// XXX: Can't use the generated client to create Foo resources.
+	// It fails with:
+	//	integration_test.go:92: error creating example foo: object *v1alpha1.Foo does not implement the protobuf marshalling interface and cannot be encoded to a protobuf message
```

Comment:
I think we'll need to run proto-gen in order to do this kind of test 🤔 

I'm not too sure why it's choosing protobuf as an encoding format - is this something new in HEAD of kubernetes?

Reply:
Ah - perhaps it is apiserver itself attempting to write to the fake etcd, which would store the message in protobuf format. I guess some part of apiextensions-apiserver would usually handle this for us?

Perhaps there's some way we could force JSON to be used instead otherwise 🤔  not sure on the best POA, but I don't think running proto-gen should be recommended *best practice* given it's only (potentially) required for integration tests.